Implemented `newline-last` rule

arturocuya opened the pull request:
Also considered naming this `eol-last` like eslint does, but I think "newline" is more undestandable.

philippe-elsass-deltatre commented on README.md:
```diff
@@ -185,6 +185,12 @@ Default rules:
 
 - `no-print`: prevent usage of `print` statements in code (`error | warn | info | off`)
 
+- `newline-last`: enforces at least one newline (or absence thereof) at the end of non-empty files
```
I think following ESlint's `eol-last` name would be sensible, but I'm ok with that

philippe-elsass-deltatre commented on src/plugins/codeStyle/diagnosticMessages.ts:
```diff
@@ -14,7 +14,9 @@ export enum CodeStyleError {
     AACommaFound = 'LINT3013',
-    AACommaMissing = 'LINT3014'
+    AACommaMissing = 'LINT3014',
+    NewlineLast = 'LINT3015',
```
Maybe `EolLastFound/EolLastMissing` would better follow the rest?

philippe-elsass-deltatre commented on src/plugins/codeStyle/index.ts:
```diff
@@ -35,6 +35,21 @@ export default class CodeStyle {
+
+        if (validateNewlineLast &&
+            // Do not validate empty files
+            file.parser.tokens.filter(t => t.kind !== TokenKind.Newline && t.kind !== TokenKind.Eof).length > 0) {
```
`tokens.filter` will process the entire file's tokens (creating a new array with it) and seems like a huge waste.
I'd recommend a better logic walking from the end.

philippe-elsass-deltatre replied:
I don't see the automatic fix so I guess it's WIP - please mark the PR as DRAFT in this case.

arturocuya replied:
Fixed by 95a4f6e

philippe-elsass-deltatre reviewed (changes requested):
Very nice contribution!

arturocuya commented:
@philippe-elsass-deltatre Update:
- Renamed rule to `eol-last`
- Added automatic fix

TwitchBronBron commented on src/plugins/codeStyle/styleFixes.ts:
```diff
@@ -142,3 +147,21 @@ function replaceFunctionTokens(diagnostic: BsDiagnostic, token: string) {
+    return {
+        diagnostic,
+        changes: [
+            insertText(diagnostic.range.end, platform.toString() === 'win32' ? '\r\n' : '\n')
```
Rather than assuming the line ending, you should just look at the line ending from the previous line. Sometimes projects will use \n on windows too.

elsassph commented on README.md:
```diff
@@ -190,6 +190,11 @@ Default rules:
 - `no-todo`: prevent usage of `todo` comments in code (`error | warn | info | off`)
 
 - `todo-pattern`: string regex pattern to determine what a TODO is. default is `TODO|todo|FIXME` (**do not** include surrounding `/` characters).
+- `eol-last`: enforces at least one newline (or absence thereof) at the end of non-empty files
```
Blank line for consistency

arturocuya replied:
Done by 95a4f6e

elsassph commented on src/plugins/codeStyle/index.ts:
```diff
@@ -36,6 +36,30 @@ export default class CodeStyle {
+
+        // Check if the file is going backwards from the last token.
+        let isFileEmpty = true;
+        for (let i = file.parser.tokens.length - 1; i >= 0; i--) {
```
Use a `const { tokens } = file.parser;` once to make code less verbose and more efficient.

arturocuya replied:
Done by 95a4f6e

arturocuya commented:
Update:
- As Bron suggested, if possible don't assume line ending for fixing. Use the last line ending instead, except when the file is a valid one-liner that needs to be fixed.
- Cleaned up code for checking if file is empty.
- Simplified how temp files are created in tests.

arturocuya commented:
Update: Rebase from master after another PR was merged.
